contracts/governance/src/governance.rs

```diff
@@ -222,52 +244,28 @@ impl GovernanceContract {
         signature: Signature,
         batch: Vec<Transfer>,
     ) -> Result<(), Error> {
-        self.validate_seed(
+        self.assert_seed(
             iter::once([seed, BlsScalar::from(TX_TRANSFER as u64)])
                 .flatten()
                 .chain(batch.iter().flat_map(Transfer::as_scalars))
                 .collect(),
             signature,
         )?;
-        self.check_pause()?;
+        self.assert_pause()?;
 
         for Transfer {
             from, to, amount, ..
         } in batch
         {
-            self.is_allowed(&from)?;
-
-            let mut base = self.balances.get(&from)?.copied().unwrap_or(0);
+            self.assert_address(&from)?;
 
-            if base < amount {
-                let remaining = amount - base;
+            let remaining = self.balance_sub(&from, amount);
 
-                self.mint(seed, signature, from, remaining)?;
-
-                base = 0;
-            } else {
-                base -= amount;
-            }
-
-            let target = self
-                .balances
-                .get(&to)?
-                .copied()
-                .unwrap_or(0)
-                .checked_add(amount)
-                .ok_or(Error::BalanceOverflow)?;
-
-            if base == 0 {
-                self.balances.remove(&from)?;
-            } else {
-                self.balances.insert(from, base)?;
+            if remaining > 0 {
+                self.mint(seed, signature, from, remaining)?
```

Comment:
As per pairing with @ZER0 we should review this in order to support all scenario